src/core/tsi/ssl/key_logging/ssl_key_logging.cc

```diff
@@ -0,0 +1,158 @@
+
+  if (fd_ == nullptr || session_keys_info.empty()) return;
+
+  switch (tsi_tls_key_log_config.key_logging_format) {
+    case GRPC_TLS_KEY_LOG_FORMAT_NSS:
+      // Append to key log file under lock
+      bool err;
+
+      err = (fwrite(session_keys_info.c_str(), sizeof(char),
+                    session_keys_info.length(),
+                    fd_) < session_keys_info.length());
+
+      // Append new-line character as well. Doing a separate syscall to avoid
+      // dynamic memory allocation and modification of passed session_keys_info
+      err |= (fwrite("\n", sizeof(char), 1, fd_) != 1);
+
+      if (err) {
+        grpc_error_handle error = GRPC_OS_ERROR(errno, "fwrite");
+        gpr_log(GPR_ERROR, "Error Appending to TLS Keylog file: %s",
+                grpc_error_std_string(error).c_str());
+        fclose(fd_);
+        fd_ = nullptr;  // disable future attempts to write to this file
+      } else {
+        fflush(fd_);
+      }
+      break;
+
+    // For future extensions to support more/custom logging formats
+    default:
+      break;
+  }
+}
+
+grpc_core::RefCountedPtr<TlsKeyLogger>
+TlsKeyLoggerRegistry::CreateTlsKeyLogger(
+    const TsiTlsKeyLoggerConfig& tsi_tls_key_log_config) {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000 && !defined(LIBRESSL_VERSION_NUMBER)
+  if (!g_tls_key_logger_registry_initialized.load()) {
+    return nullptr;
+  }
+  // To control parallel access to registry
+  grpc_core::MutexLock lock(g_tls_key_logger_registry_mu);
+  if (tsi_tls_key_log_config.tls_key_log_file_path.empty()) return nullptr;
+
+  // Check if a TlsKeyLogFileWriter instance already exists for the specified
+  // file path.
+  auto it = g_tls_key_log_file_writer_map.find(
+      tsi_tls_key_log_config.tls_key_log_file_path);
+
+  if (it == g_tls_key_log_file_writer_map.end()) {
+    // Create a new TlsKeyLogFileWriter instance
+    auto new_tls_key_log_file_writer =
+        grpc_core::MakeRefCounted<TlsKeyLogFileWriter>(
+                                  tsi_tls_key_log_config.tls_key_log_file_path);
+
+    g_tls_key_log_file_writer_map.insert(
+        std::pair<std::string, TlsKeyLogFileWriter*>(
+            tsi_tls_key_log_config.tls_key_log_file_path,
+            new_tls_key_log_file_writer.get()));
+
+    // The key logger becomes an owner of the key log file writer
+    // instance.
+    auto new_tls_key_logger =
+        grpc_core::MakeRefCounted<TlsKeyLogger>(
+            tsi_tls_key_log_config, std::move(new_tls_key_log_file_writer));
+    return new_tls_key_logger;
+  }
+
+  // The key logger also becomes an owner of the key log file writer
+  // instance.
+  auto new_tls_key_logger =
+      grpc_core::MakeRefCounted<TlsKeyLogger>(tsi_tls_key_log_config,
+                                              it->second->Ref());
+  return new_tls_key_logger;
+#endif
+  return nullptr;
+}
+
+void TlsKeyLoggerRegistry::Init() {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000 && !defined(LIBRESSL_VERSION_NUMBER)
```

Comment:
Why do we need macros here as well? It seems the code is not relevant to the OpenSSL...

Reply:
This function is removed in the latest commit.